puresnmp/x690/util.py

```diff
@@ -75,7 +90,7 @@ def from_bytes(data: Union[int, bytes]) -> "TypeInfo":
         instance._raw_value = data
         return instance
 
-    def __bytes__(self):
+    def to_bytes(self):
```

Comment:
Considering that at some point in the future (once asyncio hits) the library will no longer work in Python 2 I think it would be prudent to cut down on the size of the diff. One way to do this is revert the replacement of `__bytes__` with `to_bytes` and introduce a new `bytes()` function for Python2. In a similar way you did with the `int_from_bytes` function. This function could simply delegate the call to `__bytes__` of the object. This way the code would not need to be modified on so many places.

Something along the lines of:

    def bytes(obj):
        return obj.__bytes__()

Reply:
out of all the snmp libraries I reviewed, yours was the only one that ticked all of my boxes, except the fact that it didn't support py2 (yet). I was having a real problem with inconsistent behavior w/r/t MIB files, problems supporting the Windows environment, etc.

The problem with py3 for me is that I have a need to drop executable binaries on client servers that are being rented and the vendor they are renting from do not allow them to run installers for fear of introducing incompatibilities with the vendor's software.

Building python py3 executables with pyinstaller on windows 10 currently requires installation of the Windows UCRT on the destination machine.

Also, I use wxPython extensively and only the phoenix version is supported on Python 3, but phoenix is still very unstable.

For these reasons I am unfortunately, sadly, and painfully stuck with py2, but it pays the bills if you know what I mean.

I haven't looked into asyncio much but when I have, I've had a hard time understanding why people love it so much. Would you mind sharing with me why you want to introduce asyncio ( I'd love to better understand it's appeal ), and can I help you come up with a way of doing it that doesn't break py2 compatibility? I could always fork your project for what I'm usually your library for, but I'd rather be able to take advantage of future bug fixes.

Reply:
> I was having a real problem with inconsistent behavior w/r/t MIB files, problems supporting the Windows environment, etc.

Exactly why I started this project

> Would you mind sharing with me why you want to introduce asyncio 

Well, for starters this project was my sandbox for learning `asyncio`. I wanted to get it working first, and then introduce `asyncio`. I have not gotten around to that yet due to other projects having priority. But one day I will. But there is a chance that it can be done in a compatible manner. I will have to see. The biggest problem is that `asyncio` introduced the new `async`/`away` syntax which does not exist in Python 2. So it will break Python 2 in a fundamental manner. Maybe there are some ways around it. I'll have to see.

But it may take some time until I get around to that, so you are still good 😉 and a fork is not (yet) necessary, and is also something I would like to avoid if possible.

A simple way to explain the appeal to `asyncio` is that it allows you to easily tell the runtime: "Listen, I'm going to do I/O now, so you can do something else while I wait for a response". In essence it is *similar* (but definitely not the same) as parallelisation (threading & co), but it is easier as you don't have to deal with locks. One way of thinking about it is: Using parallelisation, context-switches are imposed by the kernel and you never know when they happen. Using `asyncio` *you* decide when to switch.

It is a really interesting model which is *perfect* for a network library like `puresnmp`. And many applications using it could really benefit from this. Alas, it is Python3 only.

Reply:
I wonder if there's a way we could introduce an opt-in asyncio wrapper in puresnmp that py3 clients can use while not introducing the breaking syntax in the core snmp logic, kinda like how you explicitly tell people that MIB processing is not within scope of the project?